### DISCUSSION TITLE
tutorials/lift-off-part2/04-implementing-our-restdatasource

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#28] [course feedback]
# tutorials/lift-off-part2/04-implementing-our-restdatasource

https://www.apollographql.com/tutorials/lift-off-part2/04-implementing-our-restdatasource

**Comment by Oceanestars** [comment]
Hi! Quick question, the "Implementing our restdatasource" section, is more if we already have an API setup and we are trying to integrate graphql to it . But we could go without it? It feels slightly redundant to have to set these getRequest when we do graphql queries.  But since a database can be a datasource we can just use/only have graphql api as our middle  part between frontend and db?
Tl;DR we don't need a rest API to get our data to the frontend with graphQL?

**Reply by mabuyo** [reply]
Hey @Oceanestars! Yes, a database can be a data source, and you do not need to use a REST API as a data source. Your GraphQL API sits in between your frontend and data sources. In this tutorial, we're using only one data source (the REST API), but you could have multiple! When our schema starts to grow, potentially using different data sources, we can see the benefits of GraphQL shine. 

But even with just the one data source we're using in the tutorial, there are still benefits! For example, the data fetching and transforming logic lives in the GraphQL server-side. This means the client can send _one clean_ query asking for exactly what it needs, no more, no less, and then render it for the user. If you can imagine without the GraphQL API, the client would have had to make a call to the REST API for `/tracks`, then make separate calls for the `/author/:id` information, and figure out how to stitch these responses together to make sense for the UI. 

Let me know if you have any additional questions or if I can clarify anything else for you!

**Reply by CMH7** [reply]
This! Exactly! This is where I am confused. Good point. But the thing is that can I really make a REST API inside of GraphQL or Apollo server? ugh I'm still confused. I don't like creating new server for the REST API.

**Reply by itsramiel** [reply]
You definitely do not need a rest api. With a typical greenfield project, you most likely will have a database as your only datasource.

The idea of having a rest datasource as used here is really more about 3rd party apis or if you already have a rest api and want to utilize it

**Reply by Theedon** [reply]
@CMH7 you do not need to have a REST API but they are so common that you would probably have to work with them on already existing projects and public APIs, the advantage with GraphQL is that you can use it as a layer in between your front-end and that RestAPI. It goes in this flow ```Frontend => GraphQL => RestAPI(optional) => Database``` 
GraphQL's biggest advantage is being able to fetch only the data you need (RestAPIs lack that ability).
Summary is that you can use GraphQL either directly with your database or even with other third party APIs(simultaneously).

**Comment by MohamedAboMousallam** [comment]
can you make a video of doing it with a functional approach?

**Reply by mabuyo** [reply]
Hey @MohamedAboMousallam, I'm not sure what you mean, could you elaborate what "functional approach" means?

**Reply by attilacsanyi** [reply]
@Mohamed as `RESTDataSource` is an abstract class afaik so you have to use `class` approach, where you can extend the RESTDataSource base class. Not sure you can achieve the same extension with just functions, but @Mabuyo let me know if there is an alternative approach :)

Thanks you all.

**Comment by Hydrock** [comment]
// Here's the solution (your code is commented out after):
class SpaceCatsAPI extends RESTDataSource {
  baseURL = "https://fake-spacecats-rest-api.cat/";
}
/* your code:
class SpaceCatsAPI extends RESTDataSource {
  constructor() {
    this.baseURL = 'https://fake-spacecats-rest-api.cat/';
  }
}
*/


but in task specified literally - It should have a constructor method.

**Comment by Hydrock** [comment]
In our class we use this.get - and yes - we inherited this method from RESTDataSource. But the article doesn't say anything about it - it says that there are some assistants

"The RESTDataSource class provides helper methods for HTTP requests. "

You shoug you should imperatively tell about the get method inherited from class RESTDataSource

**Comment by Vizigal** [comment]
warning!!!!

Inside the getAuthor method the character is 96 from the ASCII table

getAuthor(authorId) {
    return this.get(`author/${authorId}`);
  }

**Reply by repiemm** [reply]
If you are on a mac, this can be done with Option + }

**Comment by acechvz** [comment]
While solving the last code challenge, I noticed that using single quotes (') causes an error. It would be cool to either add support for it or notify the user that double quotes (") + backticks (`) are the only ones allowed.

**Reply by acechvz** [reply]
Here's the example to be more specific :D
```typescript
  getSpaceCats() {
    return this.get('spacecats'); // Causes the test to fail
  }
```

**Reply by mabuyo** [reply]
Ahh thanks for catching that @acechvz! I'll look into it.

UPDATE: Fixed!

**Comment by ThomasDos** [comment]
Hello,
In the last exercise, using encodeURIComponent caused the test to fail, which is quite unfortunate considering you had recommended it just a few lines earlier.

**Reply by mabuyo** [reply]
Hey there @ThomasDos, thanks for letting us know. The use of `encodeURIComponent` is meant to be used for your own project. We wanted to keep the code challenge simple!

**Comment by vikyw89** [comment]
why did it fail when I use
```tsx
getSpaceCats = () => {
    return this.get(`spacecats`)
}
```

**Reply by lizhennessy** [reply]
Hi there @vikyw89 - that appears to work for me. What is the error that you see when running the code challenge?

**Comment by NguyenDa18** [comment]
Great lesson!
Just a nitpick:
`Below our getTracksForHome method:`

should be

`Below our getAuthor method:`

**Reply by mabuyo** [reply]
Hey there @NguyenDa18 - apologies for the confusion. By `Below our getTracksForHome method:`, we meant that the code snippet that follows (which is the `getAuthor` method) should be inserted in the lines following the `getTracksForHome` method (ie _below_ the method). The wording is correct, but I can see how it can be misleading. Thanks for the feedback!

**Comment by SergiiBurgazlieiev** [comment]
@mabuyo I get a task error in the following GET method
```
 getMissions(catId:string){
    return this.get(`spacecats/${catId}/missions`);
  }
```
ERROR!: catId should be of type string

I implicitly assign to catId type of string. I have no idea why this task fails for me. 

Thank you!

**Reply by SergiiBurgazlieiev** [reply]
Resolved - I was needed to add space.

**Comment by JonathanCrider** [comment]
Hey just a thought: the code challenge should support utilizing encodeURIComponent, since it's discussed in the lesson. Always a good practice to be proactive about security :)

**Reply by mabuyo** [reply]
Hey there @JonathanCrider, thanks for your feedback! We wanted to keep the code challenge simple, but encourage the use of `encodeURIComponent` for your own projects.

**Reply by JonathanCrider** [reply]
No worries, it's mentioned but if you use it in the challenge that requirement fails, just thought it would be cool to support it as valid as well, just as a bonus. So either way would pass.

**Comment by joetidee** [comment]
Any reason why the getAuthor method is not abstracted into a AuthorAPI class?

**Reply by mabuyo** [reply]
For simplicity in the tutorial, so we're only managing 1 file. You can definitely extract out the author-related fetches though!
